### PR TITLE
Additions for an embedded GP system

### PR DESCRIPTION
**Pull request by darioizzo**

### CODE REVIEW

File: include/lambdifier/binary_operator.hpp [hunk 1]
```diff
@@ -2,6 +2,7 @@
 #define LAMBDIFIER_BINARY_OPERATOR_HPP
 
 #include <string>
+#include <unordered_map>
```
Comment (bluescarni) [inline include/lambdifier/binary_operator.hpp:7]: Missing ``#include <vector>``.

File: include/lambdifier/binary_operator.hpp [hunk 2]
```diff
@@ -26,11 +27,16 @@ class LAMBDIFIER_DLL_PUBLIC binary_operator
     // Getters/setters.
     const expression &get_lhs() const;
     const expression &get_rhs() const;
+    expression &access_lhs();
+    expression &access_rhs();
     void set_lhs(expression);
     void set_rhs(expression);
+    char get_op() const {return op;};
```
Comment (bluescarni) [inline include/lambdifier/binary_operator.hpp]: Is this ``get_op()`` still necessary?

File: include/lambdifier/expression.hpp [hunk 3]
```diff
@@ -124,6 +140,13 @@ class LAMBDIFIER_DLL_PUBLIC expression
         return p == nullptr ? nullptr : &(p->m_value);
     }
 
+    template <typename T>
```
Comment (bluescarni) [inline include/lambdifier/expression.hpp:143]: In pagmo we have the same name, ``extract()``, also for the unsafe version. Can you rename this?

File: include/lambdifier/function_call.hpp [hunk 4]
```diff
@@ -19,7 +19,10 @@ class LAMBDIFIER_DLL_PUBLIC function_call
 {
 public:
     bool disable_verify = false;
-
+    using eval_t = std::function<double(const std::vector<expression> &, std::unordered_map<std::string, double> &)>;
```
Comment (bluescarni) [inline include/lambdifier/function_call.hpp:23]: Missing ``#include <unordered_map>``.

File: src/binary_operator.cpp [hunk 5]
```diff
@@ -35,6 +36,16 @@ const expression &binary_operator::get_rhs() const
     return rhs;
 }
 
+expression &binary_operator::access_lhs()
```
Comment (bluescarni) [inline src/binary_operator.cpp:39]: Missing ``#include <vector>``, ``<functional>``, ``<algorithm>``.

File: src/math_functions.cpp [hunk 6]
```diff
@@ -22,6 +22,28 @@ expression sin(expression e)
     function_call fc{"llvm.sin", std::move(args)};
     fc.set_display_name("sin");
     fc.set_type(function_call::type::builtin);
+    fc.set_eval_f([](const std::vector<expression> &args, std::unordered_map<std::string, double> &v) {
```
Comment (bluescarni) [inline src/math_functions.cpp:26]: Missing ``#include <unordered_map>``.

File: include/lambdifier/number.hpp [hunk 7]
```diff
@@ -27,6 +27,9 @@ class LAMBDIFIER_DLL_PUBLIC number
 
     llvm::Value *codegen(llvm_state &) const;
     std::string to_string() const;
+    double evaluate(std::unordered_map<std::string, double> &) const;
```
Comment (bluescarni) [inline include/lambdifier/number.hpp:32]: Missing ``#include <vector>``, ``unordered_map``.

File: include/lambdifier/variable.hpp [hunk 8]
```diff
@@ -29,6 +29,9 @@ class LAMBDIFIER_DLL_PUBLIC variable
 
     llvm::Value *codegen(llvm_state &) const;
     std::string to_string() const;
+    double evaluate(std::unordered_map<std::string, double> &) const;
```
Comment (bluescarni) [inline include/lambdifier/variable.hpp:32]: Missing ``#include <vector>``, ``unordered_map``.

File: src/number.cpp [hunk 9]
```diff
@@ -1,4 +1,5 @@
 #include <string>
+#include <unordered_map>
```
Comment (bluescarni) [inline src/number.cpp:2]: Missing ``#include <vector>``

File: src/variable.cpp [hunk 10]
```diff
@@ -1,6 +1,7 @@
 #include <cstddef>
 #include <stdexcept>
 #include <string>
+#include <unordered_map>
```
Comment (bluescarni) [inline src/variable.cpp:4]: Missing ``#include <vector>``.

File: include/lambdifier/binary_operator.hpp (hunk 2 above)
Comment (darioizzo) [reply include/lambdifier/binary_operator.hpp]: I need it for the autodiff (for the moment its outside the class expression). The chain rules are different according to the operator.

Comment (darioizzo) [reply include/lambdifier/binary_operator.hpp]: I will remove at a later stage if not neede.

File: include/lambdifier/expression.hpp (hunk 3 above)
Comment (darioizzo) [reply include/lambdifier/expression.hpp:143]: sure. I am always amazed how the compiler picks up the correct one :)

**Pull request merged**